zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt

```diff
@@ -136,114 +131,124 @@ class ZiplineLoader internal constructor(
+   *     may emit periodically to trigger polling. It should also emit for loading triggers like
+   *     app launch, app foregrounding, and network connectivity changed.
+   * @param oldestCodeToLoad the age of the oldest code that this function should load. Use this to
+   *     avoid loading stale code from the cache or embedded file system after a network failure. If
+   *     no code is eligible to load the returned flow will not emit.
    */
-  private val fetchers = run {
-    val result = mutableListOf<Fetcher>()
-    val httpFetcher = HttpFetcher(httpClient, eventListener)
-    if (embeddedDir != null && embeddedFileSystem != null) {
-      result += FsEmbeddedFetcher(
-        embeddedDir = embeddedDir,
-        embeddedFileSystem = embeddedFileSystem,
-      )
-    }
-    if (cache != null) {
-      result += FsCachingFetcher(
-        cache = cache,
-        delegate = httpFetcher,
-      )
-    } else {
-      result += httpFetcher
+  suspend fun load(
+    applicationName: String,
+    manifestUrlFlow: Flow<String>,
+    oldestCodeToLoad: Duration = Duration.INFINITE,
+    initializer: (Zipline) -> Unit,
+  ): Flow<Zipline> {
+    return flow {
+      var first = true
+      var previousManifest: ZiplineManifest? = null
+
+      manifestUrlFlow.collect { manifestUrl ->
+        // Each time a manifest URL is emitted, download and initialize a Zipline for that URL.
+        //  - skip if the manifest hasn't changed from the previous load
+        //  - pin the application if the load succeeded; unpin if it failed.
+        withLifecycleEvents(applicationName, manifestUrl) {
+          val networkManifest = fetchManifestFromNetwork(applicationName, manifestUrl)
+          if (networkManifest.manifest == previousManifest) return@withLifecycleEvents // Unchanged.
+
+          try {
+            val networkZipline = loadFromManifest(applicationName, networkManifest, initializer)
+            cachingFetcher?.pin(applicationName, networkManifest) // Pin after successful loads.
+            emit(networkZipline)
+            previousManifest = networkManifest.manifest
+          } catch (e: Exception) {
+            cachingFetcher?.unpin(applicationName, networkManifest) // Unpin after failed loads.
+            throw e
+          }
+        }
+
+        // If network loading failed (due to network error, or bad code), attempt to load from the
+        // cache or embedded file system. This doesn't update pins!
+        if (previousManifest == null && first) {
+          first = false
+          val localManifest = loadCachedOrEmbeddedManifest(applicationName) ?: return@collect
+          withLifecycleEvents(applicationName, manifestUrl = null) {
+            val localZipline = loadFromManifest(applicationName, localManifest, initializer)
+            emit(localZipline)
+            previousManifest = localManifest.manifest
+          }
+        }
+      }
     }
-    result
   }
 
-  suspend fun loadOrFallBack(
```

Comment:
Previously production code was expecting to call this, which left it in a bad state when it went to load a 2nd time – how to avoid loading the same network code again?

With everything built on flows always its a lot simpler – the flow takes care of not emitting when it shouldn’t.